Fix Bugs in Sparse Communication

lroberts36 opened the pull request:
## PR Summary

This fixes a couple of bugs in sparse communication:

1. Because the field `BndInfo::allocated` was being repurposed to refer to the block being allocated for send `BndInfo` and to refer to the buffer being allocated for receive `BndInfo`, in some cases boundary prolongation would not occur when it should for sparse variables. Through a chain of events, this could show up as a dependence of results on the number of ranks that sparse problems were run on (although the differences were of order the sparse threshold). 
2. In thinking through this, I also realized that for sparse fields splitting the boundary communication into local and nonlocal is not safe, since a field can be allocated by one set of boundary communication after the other boundary communication has finished. This may not have been a real issue to date, as we have always assumed the default value to fill sparse fields with is zero, but if the default value was non-zero it could (maybe?) cause an issue.
3. The `dealloc_count` of fields was not passed during different rank, same-to-same communication during AMR. This meant that the step at which a particular block was deallocated could vary with the number of ranks.

## PR Checklist

<!-- Note that some of these check boxes may not apply to all pull requests -->

- [x] Code passes cpplint
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md
- [x] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.

lroberts36 committed:
Pass dealloc count when sending same to same

lroberts36 committed:
Almost working, but differs on step 301 after remesh

lroberts36 committed:
Add deletion check

lroberts36 committed:
Fix a bunch of bugs with a couple changes

lroberts36 committed:
Format, lint, changelog

lroberts36 committed:
Remove debugging MPI Barrier

Yurlungur commented on src/bvals/comms/bnd_info.cpp:
```diff
@@ -303,17 +306,6 @@ BndInfo BndInfo::GetSetBndInfo(std::shared_ptr<MeshBlock> pmb, const NeighborBlo
   }
 
-  if (buf_state == BufferState::received) {
-    // With control variables, we can end up in a state where a
-    // variable that is not receiving null data is unallocated.
-    // for allocated to be set, the buffer must be sending non-null
-    // data and the receiving variable must be allocated
```
Is this comment still relevant?

lroberts36 replied:
It is still relevant but I think this is not the right place to have it.

Yurlungur replied:
just noticed it's not added anywhere else. If relevant maybe stick it somewhere else in the source?

lroberts36 replied:
Thinking about this a little more, this is naturally taken care of now that I disambiguated the `BndInfo::allocated`. The comment probably doesn't need to be anywhere.

Yurlungur replied:
cool :+1:

Yurlungur commented on src/bvals/comms/boundary_communication.cpp:
```diff
@@ -252,16 +252,6 @@ TaskStatus SetBounds(std::shared_ptr<MeshData<Real>> &md) {
-    }
-
-    // Prolongate from coarse buffer
-    refinement::Prolongate(resolved_packages, cache, pmb->cellbounds, pmb->c_cellbounds);
-    refinement::ProlongateInternal(resolved_packages, cache, pmb->cellbounds,
-                                   pmb->c_cellbounds);
```
I guess these need to be split out into tasks again because of the issue with local vs nonlocal send/receive?

lroberts36 replied:
Yeah, you can't prolongate until you have received all data. As it was before, I think it would be safe with local and nonlocal, but you would prolongate twice and only the second time through would be correct. I think it is just easier to think through and less error prone to keep these separate.

Yurlungur replied:
:+1:

Yurlungur commented on src/mesh/mesh.cpp:
```diff
@@ -1086,24 +1086,53 @@ void Mesh::Initialize(bool init_problem, ParameterInput *pin, ApplicationInput *
 
     // send FillGhost variables
+    bool can_delete;
+    do {
```
Should these do-while loops have some upper iteration count in case something goes really wrong?

lroberts36 replied:
Yeah, probably

Yurlungur commented on src/mesh/mesh.cpp:
```diff
@@ -1115,18 +1144,23 @@ void Mesh::Initialize(bool init_problem, ParameterInput *pin, ApplicationInput *
+    for (int i = 0; i < nmb; ++i) {
+      auto &mbd = block_list[i]->meshblock_data.Get();
+      // Call MeshBlockData based FillDerived functions
+      Update::FillDerived(mbd.get());
```
Potential merge conflict with my precomm fill derived PR. Just need to make sure git doesn't put the precomm stuff below the boundaries/comms

lroberts36 replied:
It looks like it merged to the correct place

Yurlungur reviewed (approved):
Sounds like this was a tricky one to track down. Thanks for chasing through it!

lroberts36 committed:
Add maximum number of iterations for deletion check

lroberts36 committed:
Merge branch 'develop' into lroberts36/fix-dealloc-count-bug

pdmullen commented on src/bvals/comms/boundary_communication.cpp:
```diff
@@ -310,38 +300,28 @@ TaskStatus ApplyCoarseBoundaryConditions(std::shared_ptr<MeshData<Real>> &md) {
+  TaskStatus stat = TaskStatus::complete;
+  for (int block = 0; block < md->NumBlocks(); ++block) {
+    auto bstat = ApplyBoundaryConditionsOnCoarseOrFine(md->GetBlockData(block), false);
+    // if (bstat != TaskStatus::complete) stat = bstat;
```
Why is this commented out?

lroberts36 replied:
Oh, I should probably just remove that. `ApplyBoundaryConditionsOnCoarseOrFine` always returns `TaskStatus::complete`.

lroberts36 replied:
done

pdmullen commented on src/mesh/mesh.cpp:
```diff
@@ -1096,24 +1096,58 @@ void Mesh::Initialize(bool init_problem, ParameterInput *pin, ApplicationInput *
+          }
+        }
+      }
+    } while (!all_sent);
```
@jonahm-LANL's comments above regarding do-while termination here as well... can we just turn these into `for` loops with `break` statements?

lroberts36 replied:
A possibly better solution to this is to create and execute a `TaskList` here, which can be generated with a single call to `AddBoundaryExchangeTasks`. I think I might save this for another PR though.

pdmullen replied:
:+1:

pgrete replied:
Until this other PR is there, could we add counters?
I'm mostly concerned about burning computer time by sth going wrong without other processes knowing about it (and triggering a fail).

lroberts36 replied:
I added some simple counters, although I am not sure what a good choice for the maximum number of times through the loop is.

pdmullen commented on src/mesh/mesh.cpp:
```diff
@@ -1096,24 +1096,58 @@ void Mesh::Initialize(bool init_problem, ParameterInput *pin, ApplicationInput *
+          }
         }
       }
     } while (!all_received);
```
as above...

pdmullen reviewed (commented):
LGTM

lroberts36 committed:
Remove commented lines and clean up

lroberts36 committed:
Merge branch 'develop' into lroberts36/fix-dealloc-count-bug

lroberts36 committed:
Merge branch 'develop' into lroberts36/fix-dealloc-count-bug

lroberts36 committed:
Merge branch 'develop' into lroberts36/fix-dealloc-count-bug

pgrete commented:
@lroberts36 I was about to review but got confused by the merge conflict with `develop`.
Would you mind updating (as I'm not sure what to make of the removed commented code that we talked about in #896 with respect to the split local/nonlocal comm).

lroberts36 committed:
Merge branch 'develop' into lroberts36/fix-dealloc-count-bug

lroberts36 commented:
@pgrete: I think I just removed those comments without thinking about it. I just added the comments back. I think this is good to review now, although I am not sure what is going on with the CI. It looks like there are some things failing, but I am not sure if that is due to the issue you mentioned.

pgrete commented on src/bvals/comms/boundary_communication.cpp:
```diff
@@ -333,15 +332,13 @@ TaskID AddBoundaryExchangeTasks(TaskID dependency, TaskList &tl,
 
   // auto out = (pro_local | pro);
 
-  // TODO(LFR): Splitting up the boundary tasks while doing prolongation in one
-  //            breaks things, which I haven't completely figured out yet. To
-  //            move to the commented out code above, this needs to be fixed and
-  //            the prolongation and physical boundary conditions need to be removed
-  //            from the end of SetBounds
```
This comment is still valid given that the commented code above is also still there, isn't it?

lroberts36 replied:
Right, I will add it back

pgrete commented on src/bvals/comms/boundary_communication.cpp:
```diff
@@ -310,6 +300,15 @@ TaskStatus ApplyCoarseBoundaryConditions(std::shared_ptr<MeshData<Real>> &md) {
   return stat;
 }
 
+TaskStatus ApplyFineBoundaryConditions(std::shared_ptr<MeshData<Real>> &md) {
+  if (!md->GetMeshPointer()->multilevel) return TaskStatus::complete;
+  for (int block = 0; block < md->NumBlocks(); ++block) {
+    ApplyBoundaryConditionsOnCoarseOrFine(md->GetBlockData(block), false);
+  }
+  // ApplyBoundaryConditions is guaranteed to return complete, so this is safe
+  return TaskStatus::complete;
+}
```
Why this function (and the one above) given that `ApplyBoundaryConditionsOnCoarseOrFineMD` exists?

lroberts36 replied:
I was not aware of `ApplyBoundaryConditionsOnCoarseOrFineMD` I think. I will switch to this.

lroberts36 replied:
It looks like maybe I added `ApplyBoundaryConditionsOnCoarseOrFineMD`? Apparently my memory is completely gone

pgrete commented on src/mesh/mesh.cpp:
```diff
@@ -1084,24 +1084,58 @@ void Mesh::Initialize(bool init_problem, ParameterInput *pin, ApplicationInput *
+      }
+      test_iters++;
+    } while (!can_delete && test_iters < 1e10);
+    if (test_iters >= 1e10)
+      PARTHENON_FAIL(
+          "Too many iterations waiting to delete boundary communication buffers.");
```
`PARTHENON_REQUIRE` instead of `if ... PARTHENON_FAIL`?

lroberts36 replied:
changed.

pgrete commented on src/mesh/mesh.cpp:
```diff
@@ -1113,18 +1147,23 @@ void Mesh::Initialize(bool init_problem, ParameterInput *pin, ApplicationInput *
+        ApplyCoarseBoundaryConditions(md);
+        ProlongateBoundaries(md);
+      }
+      ApplyFineBoundaryConditions(md);
```
From `ApplyFineBoundaryConditions` above:
```c++
if (!md->GetMeshPointer()->multilevel) return TaskStatus::complete;
```

Does this logic check out? 
IIRC we also want the fine boundary conditions to be always set because those are the ones next to the active zones, isn't it?

lroberts36 replied:
Good catch, that was a left over from copy/paste of the coarse boundary conditions. I fixed this on another branch (after debugging for an embarrassingly long time), but I guess the fix never made it over here.

lroberts36 replied:
This is fixed by switching to `ApplyBoundaryConditionsOnCoarseOrFineMD`.

pgrete commented on src/outputs/parthenon_hdf5.cpp:
```diff
@@ -539,6 +539,8 @@ void PHDF5Output::WriteOutputFileImpl(Mesh *pm, ParameterInput *pin, SimTime *tm
                 vinfo.nx6 * vinfo.nx5 * vinfo.nx4 * vinfo.nx3 * vinfo.nx2 * vinfo.nx1;
           }
           memset(tmpData.data() + index, 0, varSize * sizeof(OutT));
+          for (int i = 0; i < varSize; ++i)
+            (tmpData.data() + index)[i] = std::numeric_limits<Real>::quiet_NaN();
```
If I'm reading this right, then you're overriding all the `0` from the `memset` right above, correct?
If yes, is the `memset` still required?

lroberts36 replied:
Oh, this is debugging code. I will switch back to zeros for sparse. I generally think it is more useful to plot sparse with NaNs so you can see where the sparse fields have been deallocated easily, but I think everyone agreed on filling with zeros at some point.

pgrete reviewed (commented):
Couple of minor comments and a "big" question on the fine boundaries <-> multilevel logic.

lroberts36 committed:
Respond to most of Philipp's comments

lroberts36 committed:
set maximum number of communication iterations during Mesh::Initialize

lroberts36 commented:
@pgrete: I think this is ready for a re-review and approval. You are correct about the logic being wrong for fine boundaries and multi-level, that was a copy and paste bug. This is now fixed.

pgrete committed:
Fix int datatype

pgrete reviewed (approved):
Thanks for the updates.

I just did a quick final fix (promoted the iteration counters to `int64_t` as the threshold of 1e10 was beyond the standard `int` limits) and am pulling the trigger now.

pgrete commented:
Looks like regression tests are failing, e.g.,
```
 ____Comparing on a per variable basis with tolerance 1e-12
Tolerance = 1e-12
Diff in sparse_1            
    bkji: (  71,   0,   0,   0)
    zyx: (0.000000,-0.121094,-0.246094)
    err_mag: 9.908438e-07
    f0: 5.5776e-07
    f1: 1.5486e-06
  sparse_1            : differs
Diff in sparse_0            
    bkji: ( 132,   0,   0,   0)
    zyx: (0.000000,-0.121094,0.191406)
    err_mag: 9.908438e-07
    f0: 5.5776e-07
    f1: 1.5486e-06
  sparse_0            : differs
Diff in sparse_2            
    bkji: ( 183,   0,   0,   7)
    zyx: (0.000000,0.066406,-0.191406)
    err_mag: 9.908438e-07
    f0: 5.5776e-07
    f1: 1.5486e-06
  sparse_2            : differs
Diff in sparse_3            
    bkji: ( 244,   0,   0,   7)
    zyx: (0.000000,0.066406,0.246094)
    err_mag: 9.908438e-07
    f0: 5.5776e-07
    f1: 1.5486e-06
  sparse_3            : differs
Traceback (most recent call last):
  File "/__w/parthenon/parthenon/tst/regression/run_test.py", line 204, in <module>
    main(**vars(args))
  File "/__w/parthenon/parthenon/tst/regression/run_test.py", line 93, in main
    raise TestError("Test " + test_manager.test + " failed")
__main__.TestError: Test sparse_advection failed
```
Might be related to the fine boundary conditions?

lroberts36 commented:
@pgrete: Oh, I think I know what is going on. I updated the sparse gold file with the bugs in. I think this test succeeds with the old gold files with the bug fixes in. I will make sure this is true and update.

lroberts36 committed:
Update gold standard
